Enable expressions in lists

NicoLaval opened the pull request:
Fix #506

NicoLaval committed:
Enable expressions in lists

vpinna80 commented:
Hi @NicoLaval, can you distinguish between inExpr and inExprComponent?
To avoid introducing unwanted paths in the grammar, if "lists" uses "expr" in "inNotInExpr" then "listsComponent" uses "exprComponent" in "inNotInExprComponent". 

Also I think we must change the documentation in order to clarify:
* all items in the set must be scalars
* all items in the set must be convertible to the valuedomain of the value being tested
* -or- all items in the set must belong to the same valuedomain and the value being tested must be convertible to that domain.

This is to make sure that we can determine valuedomain errors at the semantic validation step.

NicoLaval committed:
Merge branch 'master' into Fix/improve-lists

NicoLaval committed:
Update list in G4

NicoLaval commented:
Hi @vpinna80,

G4 is updated, with 2 branches, one for scalars and one for components.

I also updated the name `lists` with `list`: I think the plural didn't make sense.

Regarding the docs, please provide your updates.

Shouldn't we also add examples with components?

javihern98 commented:
I do not understand the use of listComponent, I believe the original proposal was fine if there are no ambiguities

NicoLaval commented:
> > I do not understand the use of listComponent, I believe the original proposal was fine if there are no ambiguities
> 
> Maybe @vpinna80 wants to be able to write something like:
> 
> `ds_out := ds_in [calc ind_me1 := me1 in { a, b }]` where `a` and `b` are components of `ds_in`?

Which on the other hand no longer allows us to compare a component one to constant scalars.

vpinna80 commented:
Yes, only variables which are components can be used inside the square brackets, no other.

@NicoLaval Regarding the docs, i cannot update your pull request, but you should be able to edit the doc file directly on GitHub:
https://github.com/sdmx-twg/vtl/blob/Fix/improve-lists/v2.1/docs/reference_manual/operators/Comparison%20operators/Element%20of/content.rst

javihern98 commented:
> Yes, only variables which are components can be used inside the square brackets, no other.

Just to clarify, I understand by your suggestion the in operator will use therefore:
- A set of scalars
- A list of components, if used inside a calc if the names included in the list are also components of the dataset

Seeing both use cases, it still makes no sense to have both list and listComponent. I believe it goes out of scope of the issue anyway to add the compatibility of component names inside the in operator

NicoLaval committed:
Merge branch 'master' into Fix/improve-lists

hadrienk commented:
I think adding listComponent is the only way to ensure that expr and exprComp are similar. 

Why dont you think it make sense @javihern98 ?

hadrienk commented:
> Yes, only variables which are components can be used inside the square brackets, no other.

Why do you think external variables should no be possible? 

In any case, i think this can easily be a runtime check, not a grammar check.

hadrienk commented:
> > I think adding listComponent is the only way to ensure that expr and exprComp are similar.
> > Why dont you think it make sense @javihern98 ?
> 
> Well the use of this token will we the same and have the same syntax whether we use it inside a calc or not, and we cannot use this token outside of the in operator. So to separate them in list and listComponent I find it redundant

Ah, I see. I think this boils down to the duplicated expr branches then. I think we touched upon the subject in Salamanca. The whole exprComp rule branch only exists to account for a few disparities that only apply to components. I don't remember what they were but this really be fixed. 

@NicoLaval didn't we create an issue about this?

We ended up patching the grammar in trevas to only have one expr rule.

NicoLaval commented:
> > > I think adding listComponent is the only way to ensure that expr and exprComp are similar.
> > > Why dont you think it make sense @javihern98 ?
> > 
> > 
> > Well the use of this token will we the same and have the same syntax whether we use it inside a calc or not, and we cannot use this token outside of the in operator. So to separate them in list and listComponent I find it redundant
> 
> Ah, I see. I think this boils down to the duplicated expr branches then. I think we touched upon the subject in Salamanca. The whole exprComp rule branch only exists to account for a few disparities that only apply to components. I don't remember what they were but this really be fixed.
> 
> @NicoLaval didn't we create an issue about this?
> 
> We ended up patching the grammar in trevas to only have one expr rule.

Hi @hadrienk, @javihern98,
No we don't, other TF members were not really for this change.

Beyond producing a tree twice as big, this does not allow mixing the types (I will do a PR for the Levenshtein distance during the day, I will have the same problem).

In this context, for instance, consider:

`ds_out := ds_in [calc a := me_1 in { me_2, "default" }][drop me_1, me_2];`

With `ds_in` like:
|id_1|me_1|me_2|
|-|-|-|
|1|"foo"|"foo"|
|2|"foo"|"bar"|
|3|"default"|"baz"|

To produce `ds_out`:
|id_1|me_1|
|-|-|
|1|true|
|2|false|
|3|true|

This is not valid while we keep this `expr`/`exprComponent` distinction.

However, it seems to me that defining this script makes sense.

vpinna80 commented:
Why is it not valid? I tested it with the change and it parses correctly.

```
expr:
    | left=expr op=(IN|NOT_IN)(lists|valueDomainID)                         # inNotInExpr

exprComponent:
    | left=exprComponent op=(IN|NOT_IN)(listsComponent|valueDomainID)          # inNotInExprComp

lists:
    GLPAREN  expr (COMMA expr)*  GRPAREN

listsComponent:
    GLPAREN  exprComponent (COMMA exprComponent)*  GRPAREN
```

NicoLaval commented:
> Why is it not valid? I tested it with the change and it parses correctly.
> 
> ```
> expr:
>     | left=expr op=(IN|NOT_IN)(lists|valueDomainID)                         # inNotInExpr
> 
> exprComponent:
>     | left=exprComponent op=(IN|NOT_IN)(listsComponent|valueDomainID)          # inNotInExprComp
> 
> lists:
>     GLPAREN  expr (COMMA expr)*  GRPAREN
> 
> listsComponent:
>     GLPAREN  exprComponent (COMMA exprComponent)*  GRPAREN
> ```

Syntactic validation is good, but if you resolve expr and exprComponent differently in your engine, it will bug.

And, if it doesn't bug, it's because you infer the types at runtime, and in this case, why keep the 2 branches?

NicoLaval commented:
@linardian , why closed & deleted?

linardian commented:
I fixed the error in 2.1.
Since the Levenshtein operator was not defined according to the agreed syntax, to avoid confusion I dropped the whole changes.
If you do not agree please let me know how to proceed
Best

Angelo.


Inviato con BlackBerry Work
(www.blackberry.com)


Da: Nicolas Laval ***@***.******@***.***>>
Data: sabato 19 apr 2025, 10:18 AM
A: sdmx-twg/vtl ***@***.******@***.***>>
Cc: Angelo Linardi ***@***.******@***.***>>, Mention ***@***.******@***.***>>
Oggetto: Re: [sdmx-twg/vtl] Enable expressions in lists (PR #537)

@ linardian , why closed & deleted? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned. Message ID: <sdmx-twg/vtl/pull/537/c2816725516@ github. com> NicoLaval left


@linardian<https://urldefense.com/v3/__https://github.com/linardian__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl6bvMWRMA$> , why closed & deleted?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/sdmx-twg/vtl/pull/537*issuecomment-2816725516__;Iw!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl5KSkd_BA$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AWRHAJUWDBOYU55MPDSUB6322JLMBAVCNFSM6AAAAABV424AWWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWG4ZDKNJRGY__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl44JhWuuA$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[https://avatars.githubusercontent.com/u/30119477?s=20&v=4]NicoLaval left a comment (sdmx-twg/vtl#537)<https://urldefense.com/v3/__https://github.com/sdmx-twg/vtl/pull/537*issuecomment-2816725516__;Iw!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl5KSkd_BA$>

@linardian<https://urldefense.com/v3/__https://github.com/linardian__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl6bvMWRMA$> , why closed & deleted?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/sdmx-twg/vtl/pull/537*issuecomment-2816725516__;Iw!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl5KSkd_BA$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AWRHAJUWDBOYU55MPDSUB6322JLMBAVCNFSM6AAAAABV424AWWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWG4ZDKNJRGY__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl44JhWuuA$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

** Le e-mail provenienti dalla Banca d'Italia sono trasmesse in buona fede e non comportano alcun vincolo né creano obblighi per la Banca stessa, salvo che ciò non sia espressamente previsto da un accordo scritto. Questa e-mail è confidenziale. Qualora l'avesse ricevuta per errore, La preghiamo di comunicarne via e-mail la ricezione al mittente e di distruggere il contenuto. La informiamo inoltre che l'utilizzo non autorizzato del messaggio o dei suoi allegati potrebbe costituire reato. La Banca d’Italia tratta i dati personali in linea con il Regolamento Europeo 2016/679 (GDPR). Per informazioni può consultarsi “Privacy” sul sito della Banca d’Italia. Per ulteriori richieste è possibile rivolgersi al Titolare del trattamento ***@***.***). E’ possibile anche contattare il Responsabile della protezione dei dati ***@***.*** **
Prima di stampare, pensa verde

** E-mail from Bank of Italy are sent in good faith but they are neither binding on the Bank nor to be understood as creating any obligation on its part except where provided for in a written agreement. This e-mail is confidential. If you have received it by mistake, please inform the sender by reply e-mail and delete it from your system. Please also note that the unauthorized disclosure or use of the message or any attachments could be an offence. The Bank of Italy processes personal data in line with Regulation (EU) 2016/679 (General Data Protection Regulation - GDPR). For any further information you can consult the Privacy Policy on the Bank of Italy webpage. In case of queries, please contact the Bank of Italy Data Controller ***@***.***) . You may also contact the Bank of Italy Data Protection Officer ***@***.*** . Thank you for your cooperation **
Think green before you print

NicoLaval commented:
> I fixed the error in 2.1. Since the Levenshtein operator was not defined according to the agreed syntax, to avoid confusion I dropped the whole changes. If you do not agree please let me know how to proceed Best Angelo. Inviato con BlackBerry Work ([www.blackberry.com](http://www.blackberry.com)) Da: Nicolas Laval ***@***.******@***.***>> Data: sabato 19 apr 2025, 10:18 AM A: sdmx-twg/vtl ***@***.******@***.***>> Cc: Angelo Linardi ***@***.******@***.***>>, Mention ***@***.******@***.***>> Oggetto: Re: [sdmx-twg/vtl] Enable expressions in lists (PR #537) @ linardian , why closed & deleted? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned. Message ID: <sdmx-twg/vtl/pull/537/c2816725516@ github. com> NicoLaval left @linardian<https://urldefense.com/v3/__https://github.com/linardian__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl6bvMWRMA$> , why closed & deleted? — Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/sdmx-twg/vtl/pull/537*issuecomment-2816725516__;Iw!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl5KSkd_BA$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AWRHAJUWDBOYU55MPDSUB6322JLMBAVCNFSM6AAAAABV424AWWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWG4ZDKNJRGY__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl44JhWuuA$>. You are receiving this because you were mentioned.Message ID: ***@***.***> [https://avatars.githubusercontent.com/u/30119477?s=20&v=4]NicoLaval left a comment ([sdmx-twg/vtl#537](https://github.com/sdmx-twg/vtl/pull/537))<https://urldefense.com/v3/__https://github.com/sdmx-twg/vtl/pull/537*issuecomment-2816725516__;Iw!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl5KSkd_BA$> @linardian<https://urldefense.com/v3/__https://github.com/linardian__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl6bvMWRMA$> , why closed & deleted? — Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/sdmx-twg/vtl/pull/537*issuecomment-2816725516__;Iw!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl5KSkd_BA$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AWRHAJUWDBOYU55MPDSUB6322JLMBAVCNFSM6AAAAABV424AWWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWG4ZDKNJRGY__;!!DKHwpfUEEKarIw!qxdyq2nSKozDtbt0NWa_X0A09M1au1djYxrLjM8CwQLe0TA8X57Hygs8MLyIeFbLxpvAnqM-ZZ2jAS7uhm593_u5Wl44JhWuuA$>. You are receiving this because you were mentioned.Message ID: ***@***.***> ** Le e-mail provenienti dalla Banca d'Italia sono trasmesse in buona fede e non comportano alcun vincolo né creano obblighi per la Banca stessa, salvo che ciò non sia espressamente previsto da un accordo scritto. Questa e-mail è confidenziale. Qualora l'avesse ricevuta per errore, La preghiamo di comunicarne via e-mail la ricezione al mittente e di distruggere il contenuto. La informiamo inoltre che l'utilizzo non autorizzato del messaggio o dei suoi allegati potrebbe costituire reato. La Banca d’Italia tratta i dati personali in linea con il Regolamento Europeo 2016/679 (GDPR). Per informazioni può consultarsi “Privacy” sul sito della Banca d’Italia. Per ulteriori richieste è possibile rivolgersi al Titolare del trattamento ***@***.***). E’ possibile anche contattare il Responsabile della protezione dei dati ***@***.*** ** Prima di stampare, pensa verde ** E-mail from Bank of Italy are sent in good faith but they are neither binding on the Bank nor to be understood as creating any obligation on its part except where provided for in a written agreement. This e-mail is confidential. If you have received it by mistake, please inform the sender by reply e-mail and delete it from your system. Please also note that the unauthorized disclosure or use of the message or any attachments could be an offence. The Bank of Italy processes personal data in line with Regulation (EU) 2016/679 (General Data Protection Regulation - GDPR). For any further information you can consult the Privacy Policy on the Bank of Italy webpage. In case of queries, please contact the Bank of Italy Data Controller ***@***.***) . You may also contact the Bank of Italy Data Protection Officer ***@***.*** . Thank you for your cooperation ** Think green before you print

Hi Angelo,

Just by restoring the branch, and let the discussion and code adjustments take place.

Thanks

hadrienk commented:
> I fixed the error in 2.1. Since the Levenshtein operator was not defined according to the agreed syntax, to avoid confusion I dropped the whole changes. If you do not agree please let me know how to proceed

Hi Angelo. Until a branch is "merged" it has no impact with the current version so there's no need to close/delete them.
